### ISSUE TITLE
Allow to compile an implementation against an interface that does not reside in the same directory

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/ocaml#9717]
As far as I understand it is currently it is not possible to constrain the compilation of an implementation by an interface if there is no corresponding `.mli` file in the *same directory* as the compiled `.ml` file (and whatever there may be in that `.mli` file). 

Naïvely I had expected the following to work: 
```
ocamlc -c iface/a.mli -o build/a.cmi
ocamlc -I build -c impl/a.ml -o build/a.cmo
```
But it does not: 
```
> mkdir -p iface
> mkdir -p impl
> mkdir -p build
> echo 'val x : int' > iface/a.mli
> echo 'let x = 3 let y = 4' > impl/a.ml
> ocamlc -c iface/a.mli -o build/a.cmi
> ocamlobjinfo build/a.cmi | grep 'A' | tail -1
	f5fcec1265dd27139cb0c9ff5ae77fae	A
> ocamlc -I build -c impl/a.ml -o build/a.cmo
> ocamlobjinfo build/a.cmi | grep 'A' | tail -1
	75202fedaa6dc76230d68e37bde70d21	A
> ocamlobjinfo build/a.cmo | grep 'A' | tail -1
	75202fedaa6dc76230d68e37bde70d21	A
> touch impl/a.mli # We just need a file here, contents is irrelevant
> ocamlc -c iface/a.mli -o build/a.cmi        
> ocamlobjinfo build/a.cmi | grep 'A' | tail -1
	f5fcec1265dd27139cb0c9ff5ae77fae	A
> ocamlc -I build -c impl/a.ml -o build/a.cmo  
> ocamlobjinfo build/a.cmi | grep 'A' | tail -1
	f5fcec1265dd27139cb0c9ff5ae77fae	A
> ocamlobjinfo build/a.cmo | grep 'A' | tail -1
	f5fcec1265dd27139cb0c9ff5ae77fae	A
```

For some reason it seems the `.cmi` lookup is guarded by a test for a `.mli` in the *same directory* [here](https://github.com/ocaml/ocaml/blob/1fb6a464a8c8952763bdb2e04d05637bce55c3e6/typing/typemod.ml#L2652-L2654). I'm not sure what the rationale is. The current code also leads to suprising context sensitive cli interactions: 

```
> touch a.ml    
> ocamlc -c a.ml 
> rm a.cmi
> touch a.mli   
> ocamlc -c a.ml
File "a.ml", line 1:
Error: Could not find the .cmi file for interface a.mli.
```
(I suppose this tries to help end users making correct sequences of compilation but a warning would maybe have been more appropriate here). 

I wonder if it would be possible to change this or add a flag that prevents this. One (mostly) backward compatible way of dealing with the problem would be to maybe look for the `.mli` in the include path rather than in the same directory this would allow to support the above source structure by compiling with:

```
ocamlc -I intf -I build -c impl/a.ml -o build/a.cmo
```

**Comment by shindere** [comment]
Yes, this issue makes a lot of sense to me in the context of
out-of-source-tree builds. As soon as the `.ml` file will be generated,
it will have to be in the build rather than source tree whereas the .mli
file will most likely not need to be generated and will thus be part
of the source tree.

**Comment by elfring** [comment]
I became also curious how the support for out-of-source-tree builds will evolve further.

**Comment by damiendoligez** [comment]
>For some reason it seems the .cmi lookup is guarded by a test for a .mli in the same directory here. I'm not sure what the rationale is.

The problem is that the compilation of the .ml file also produces a .cmi, *if and only if* the .mli file is absent. Checking for the presence of the .cmi itself is not good because it could be left over from a previous compilation of the .ml. It (somewhat) makes sense to check for the presence of the .mli file right next to the .ml, since they are both source files, and strongly related.

If we add a mode where the .mli is mandatory and compiling the .ml never generates the .cmi, then I believe it becomes easy to specify and implement what you want. Or maybe just a flag to skip both the .mli test and the .cmi generation.

**Comment by shindere** [comment]
Damien Doligez (2021/05/20 02:51 -0700):
> >For some reason it seems the .cmi lookup is guarded by a test for a .mli in the same directory here. I'm not sure what the rationale is.
>
> The problem is that the compilation of the .ml file also produces a
> .cmi, *if and only if* the .mli file is absent.

This can indeed be a problem, in the current context.

> Checking for the presence of the .cmi itself is not good because it
> could be left over from a previous compilation of the .ml. It
> (somewhat) makes sense to check for the presence of the .mli file
> right next to the .ml, since they are both source files, and strongly
> related.

Not sure I agree, here. To give you just one example, `stdlib/sys.ml` is
now a configured module, meaning it is generated by `configure` from
`sys.ml.in`. At the moment it's not a big deal, but it will become when
we start to de-fusion source and build-tree because then `sys.mli` will
be somewhere in the source tree, whereas `sys.ml` could potnetially be
in a totally different directory. In such a case I expect `sys.cmi` to
be besides `sys.ml` in the build tree, rather than besides `sys.mli` in
the source tree, because both files are generated.

> If we add a mode where the .mli is mandatory and compiling the .ml
> never generates the .cmi, then I believe it becomes easy to specify
> and implement what you want. Or maybe just a flag to skip both the
> .mli test and the .cmi generation.

The second option looks better to me.

What do you thingk, @damiendoligez and @dbuenzli, and others? :-)

**Comment by damiendoligez** [comment]
> The second option looks better to me.

After thinking about it, I think you're right. In practice it means the .mli is still mandatory (because it's the source for the .cmi, which is needed) but the compiler doesn't enforce its location.

**Comment by shindere** [comment]
Damien Doligez (2021/09/10 08:17 -0700):
> > The second option looks better to me.
>
> After thinking about it, I think you're right. In practice it means
> the .mli is still mandatory (because it's the source for the .cmi,
> which is needed) but the compiler doesn't enforce its location.

So, what would a good name be for such a flag, according to you?

Would `-no-cmi` be good enough?

**Comment by dbuenzli** [comment]
> One (mostly) backward compatible way of dealing with the problem would be to maybe look for the `.mli` in the include path rather than in the same directory

Why wasn't this option considered ? I have the impression it fits better in the way the compilation toolchain works.

**Comment by shindere** [comment]
Daniel Bünzli (2021/09/16 05:50 -0700):
> > One (mostly) backward compatible way of dealing with the problem would be to maybe look for the `.mli` in the include path rather than in the same directory
> 
> Why wasn't this option considered ? I have the impression it fits
> better in the way the compilation toolchain works.

I have no opinion. I am waiting until everybody aggrees on a solutionso
that I can implement it.

**Comment by gasche** [comment]
I don't see a problem with looking for the `.mli` in the load path rather than just the `.ml` directory, and it does seem aligned with the rest of the looking semantics. There could be performance reasons to stick to the current behavior, as scanning the load path can be costly on some filesystems. But if I remember correctly, we now cache the load-path file listing, so this should in fact be cheap.

Note: in the (still-not-merged) #9056 we discuss an option `-inc foo/bar.cmi` to add a single cmi file to the load path. With my understanding of the current discussion, `ocamlc -inc iface/test.cmi test.ml` would not work, as `iface/test.mli` would not be added to the load path. Maybe we should also support `ocamlc -inc iface/test.mli -inc iface/test.cmi test.ml` in this case -- have `-inc` work for `.mli` files as well.

**Comment by nojb** [comment]
> Note: in the (still-not-merged) #9056 we discuss an option `-inc foo/bar.cmi` to add a single cmi file to the load path. With my understanding of the current discussion, `ocamlc -inc iface/test.cmi test.ml` would not work, as `iface/test.mli` would not be added to the load path. Maybe we should also support `ocamlc -inc iface/test.mli -inc iface/test.cmi test.ml` in this case -- have `-inc` work for `.mli` files as well.

Thanks for the reminder :) Is the suggestion to allow `.mli` lookup when using `-inc` but not necessarily `-I`? Up until now, the semantics of the projected `-inc` flag were defined in terms of those of `-I`, but doing this would break this symmetry. Conversely, if we were to allow using `-I` for this, we should necessarily also allow it for `-inc`.

**Comment by gasche** [comment]
My point that if we want to allow lookup of .mli in other places (eg. with `-I`), we should make sure to also support them with `-inc` otherwise `-inc other-place/foo.cmi` is not going to work -- because the corresponding mli will not be found. I think we discussed at some point which files should be `-inc`-cludeable (all? just `.cmi`? I think the consensus was on "all" anyway), that's one more.

**Comment by xavierleroy** [comment]
> I don't see a problem with looking for the .mli in the load path rather than just the .ml directory, and it does seem aligned with the rest of the looking semantics. There could be performance reasons to stick to the current behavior, as scanning the load path can be costly on some filesystems.

Compilation time is not an issue: the compilers do a great many lookups for .cmi files, anything on top of that is negligible.

As far as I can remember, the current behavior (demand that the `.mli` interface is in the same directory as the `.ml` implementation) was intended to make life easier for editors (it's trivial to switch between implementation and interface) and maybe even for users (no need to chase the `.mli` down a path when we know where the `.ml` is).

Is there a strong need for the proposed change?

Are editors ready?  (e.g. Merlin)

Are we sure we're not making the user's life more difficult?

**Comment by Octachron** [comment]
I am wondering if it would not be simpler to an option to explicitly associate a cmi to a ml file something like, `ocamlc -c a.ml:dir/a.cmi`?

**Comment by alainfrisch** [comment]
> With my understanding of the current discussion, ocamlc -inc iface/test.cmi test.ml would not work, as iface/test.mli would not be added to the load path. Maybe we should also support ocamlc -inc iface/test.mli -inc iface/test.cmi test.ml in this case -- have -inc work for .mli files as well.

I believe that `-inc` would mostly be used by clever build systems, to control the visibility of .cmi files and avoid useless copying of files, just to please the compiler.  But a clever build system also knows whether the corresponding .mli file exists (/ can be built) or not, and could pass the information to the compiler.  What about adding flags to tell the compiler explicitly whether it should (1) produce the .cmi (from the .ml file), or (2) load the .cmi (from the load path or `-inc` flags) and fail if it cannot read it?  This seems more robust than having a behavior that depends on the presence of the .mli file.

**Comment by damiendoligez** [comment]
>What about adding flags to tell the compiler explicitly whether it should (1) produce the .cmi (from the .ml file), or (2) load the .cmi (from the load path or -inc flags) and fail if it cannot read it? This seems more robust than having a behavior that depends on the presence of the .mli file.

I like this solution, it's simple and robust.

**Comment by gasche** [comment]
So we had a development meeting where we touched this issue. Here is what was said:

- These days Merlin could easily locate a `.mli` in a different place if necessary.
- If we want the feature, there is consensus on @alainfrisch's interface suggestion.
- This feature is not a priority (compared to, say, the relocatability work), so no one was committing to implement this for now.
  (I think it could also come after some other Load_path-related work, such as the `-inc` stuff we were discussing.)

**Comment by nojb** [comment]
Two thoughts:

- If we take @alainfrisch's line of reasoning further, in the case where we ask the compiler to find an existing `.cmi` file one should be able to pass the exact path on the command-line instead of having the compiler look it up in the load path (after all, this is the kind of information that would be well known to a clever build system). Or perhaps this should be a third option (so: 1) produce `.cmi` from `.ml` file, 2) use an existing `.cmi` file from the load path, 3) use an existing `.cmi` file passed explicitly on the command-line).

- Yet another option could be telling the compiler that a `.cmi` corresponding to an empty signature should be produced from a `.ml` file. This is useful for when modules are used solely for their side effects. Currently this use-case is not well supported by the compiler (see eg the second bullet point in https://discuss.ocaml.org/t/dealing-with-warning-70-missing-mli/8542). At LexiFi we have a local patch to the compiler that allows specifying the interface of a module in the `.ml` file that we use precisely for this purpose.

**Comment by alainfrisch** [comment]
One more thought :  to better support the common case of an empty signature, what about making it so the compiler accepts empty .cmi files (i.e. files with size 0) and interprets them as empty interfaces?  This means the build system could efficiently create those files itself (even before compiling the implementation, which is better for incremental compilation, at least in byte code), without the need to create first an empty .mli file and call the compiler on it.

**Comment by nojb** [comment]
Going along @Octachron's suggestion, we could have a mini "grammar" for command-line `.cmi` specification for `.ml` arguments:
```
foo.ml@path/to/bar.cmi # specifying an explicit .cmi
foo.ml@infer # infer the .cmi from .ml; fail if .mli file exists next to .ml
foo.ml@empty # build an "empty" .cmi from .ml; fail if .mli file exists next to .ml
foo.ml@find # look up .cmi file in load path
foo.ml@default # current logic depending on the presence of .mli file next to .ml (default)
```
This concerns only the lookup of `.cmi` files. Previously the lookup of `.mli` files in the load path was also mentioned, but it wasn't clear to me if this was only mentioned as a way of controlling the lookup of `.cmi` files only or if there was interest on controlling the lookup of `.mli` files on its own.

**Comment by shindere** [comment]
Not sure a "grammar" is a good approach but I understand all the
describe behaviours are interesting.

Wouldn't it be okay to use just flags?

For instance, how about a `-with-intf` or `-with-dcmi` command-line
option taking as argument the file to use as an interface? Am I correct
that such a flag owuld

 1. Address this issue
 2. Open the road to the idea of empty `.cmi` files?

**Comment by nojb** [comment]
> Wouldn't it be okay to use just flags?

Not sure how important it is in practice, but using a flag makes it less usable when passing more than one `.ml` file on the command-line at the same time (eg if you wanted to specify explicitly `.cmi` file for each `.ml` file argument).

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2021/09/30 03:02 -0700):
> > Wouldn't it be okay to use just flags?
>
> Not sure how important it is in practice, but using a flag makes it
> less usable when passing more than one `.ml` file on the command-line
> at the same time (eg if you wanted to specify explicitly `.cmi` file
> for each `.ml` file argument).

Yeah, I thought about this limitation but had the same question than
you, e.g. how often does that happen in practice that several .ml files
are compiled at the same time.

**Comment by damiendoligez** [comment]
After some brainstorming with @shindere, it seems clear that we need to support at least the following scenarios for compilation out of source:

1. The .ml and .mli are together in the source directory, and the .cmi is in the build directory
2. The .mli is in the source directory and the .ml is generated: it sits in the build directory along with the .cmi

This can be supported with two command-line options:
- `-cmi-dir <d>` look for or generate the .cmi file in `<d>` (default: same as the .ml file)
- `-mli-dir <d>` look for the .mli file in `<d>` (default: same as the .ml file)

This is orthogonal to the question of forbidding the auto-generation of the .cmi when the .mli is absent.

This is nicely similar to:
- `-odir <d>` generate the .cmo/.cmx/.cmt/... files in `<d>` (default: same as the .ml file)

**Comment by dbuenzli** [comment]
> 2\. The .mli is in the source directory and the .ml is generated: it sits in the build directory along with the .cmi

FWIW I just hit that problem this week in a project. I had to copy the `.mli` to the build directory which is annoying. 

> This can be supported with two command-line options:

Is there any strong reason not to specify the files directly (instead of directories) ? In my build systems most of the time I do have all the file paths (because of precise dependency tracking) and I spend my time trimming them to directories to pass as `-I` :-) (See #7080)

**Comment by dbuenzli** [comment]
> the following scenarios

Also I would like to add (but I think it's supported by your proposal). 

3. The `.mli` is in one source directory, the `.ml` in another one and the `.cmi` is in the build directory. 

This occurs in these cases:

a. Module level conditional compilation e.g. `m.mli` and implementations for different OCaml versions `4.10/m.ml`, `4.11/m.ml `, `4.12/m.ml`
b. Interface implemented by multiple backends, e.g. `m.mli` and implementation `os/m.ml`, `jsoo/m.ml`.

**Comment by damiendoligez** [comment]
> Is there any strong reason not to specify the files directly (instead of directories) ? In my build systems most of the time I do have all the file paths (because of precise dependency tracking) and I spend my time trimming them to directories to pass as -I :-) (See #7080)

Makes it easier to pass several source files to one compiler invocation. although I don't know if you'll see it as a *strong* reason.

**Comment by shindere** [comment]
I just opened #10981 with a proposed fix.

Comments welcome and warmly appreciated.

**Comment by damiendoligez** [comment]
Fixed by #10981.
